:arrow_up: Scala 2.13.8

cheleb opened the pull request.

cheleb commented:
I struggled and loose with the crossversion configuation :(

scottsand-db commented:
@cheleb "I struggled and loose with the crossversion configuation :("

What are you struggling with? Is some command failing? Let me know how I can help.

scottsand-db commented on standalone/src/main/scala-2.12/org/scalatest/funsuite/AnyFunSuite.scala:
```diff
@@ -0,0 +1,3 @@
+package org.scalatest.funsuite
+
+class AnyFunSuite extends org.scalatest.FunSuite
```
Hey, can you explain what's going on here?

cheleb replied:
Commit mistake...

scottsand-db commented on standalone/src/main/scala-2.12/scala/collection/parallel/CollectionConverters.scala:
```diff
@@ -0,0 +1,5 @@
+package scala.collection.parallel
+
+object CollectionConverters {
```
And here?

cheleb replied:
Mistake sorry :(

cheleb replied:
It a was a bad attempt to avoid use specific scala src folder and to duplicate code.
I've forgot to commit this removal.
Sorry for the noise.

scottsand-db commented on build.sbt:
```diff
@@ -40,7 +40,8 @@ val sparkVersion = "2.4.3"
 val parquet4sVersion = "1.2.1"
 val parquetHadoopVersion = "1.10.1"
-val scalaTestVersion = "3.0.5"
+
```
btw, let's try and avoid noise like empty lines.

scottsand-db commented on build.sbt:
```diff
@@ -661,3 +671,5 @@ lazy val flinkConnector = (project in file("flink-connector"))
   .settings(skipReleaseSettings)
   .dependsOn(standaloneCosmetic % "provided")
+
+  */
```
nit: new line at end of file.

scottsand-db commented on sql-delta-import/src/main/scala/ImportRunner.scala:
```diff
@@ -22,9 +22,9 @@ import org.rogach.scallop.{ScallopConf, ScallopOption}
  * Spark app that wraps functionality of JDBCImport and exposes configuration as command line args
  */
-object ImportRunner extends App {
+object ImportRunner  {
```
can you explain this change too?

cheleb replied:
AFAIU this is useless and moreover main is final in scala 2.13

scottsand-db commented on standalone/src/main/scala/io/delta/standalone/internal/SnapshotImpl.scala:
```diff
@@ -36,6 +37,7 @@ import io.delta.standalone.internal.logging.Logging
 import io.delta.standalone.internal.scan.{DeltaScanImpl, FilteredDeltaScanImpl}
 import io.delta.standalone.internal.util.{ConversionUtils, FileNames, JsonUtils}
 
+
```
nit: noise.

scottsand-db commented on standalone/src/main/scala/io/delta/standalone/internal/util/CaseInsensitiveMap.scala:
```diff
@@ -29,6 +29,13 @@ import java.util.Locale
+  override def updated[V1 >: T](key: String, value: V1): Map[String, V1] =
+     new CaseInsensitiveMap(originalMap.updated(key.toLowerCase(Locale.ROOT), value))
+
+
```
nit: noise

scottsand-db commented:
You will also need to update `.github/workflows/test.yaml` and update the `scala` array values to contain the latest scala version.

I also think we might have to do some MiMa magic (e.g. like [here](https://github.com/scala/sbt-scala-module/pull/59/files)) to fix our javadoc compilation...

zsxwing commented on .github/workflows/test.yaml:
```diff
@@ -6,7 +6,7 @@ jobs:
     strategy:
       matrix:
-        scala: [2.12.8, 2.11.12]
+        scala: [2.13.8, 2.12.8, 2.11.12]
```
Could you remove Scala 2.11 from all places? We want to support only two Scala minor versions.

zsxwing commented on build.sbt:
```diff
@@ -593,21 +599,21 @@ lazy val goldenTables = (project in file("golden-tables")) settings (
     name := "sql-delta-import",
     commonSettings,
-    publishArtifact := scalaBinaryVersion.value == "2.12",
+    publishArtifact := (scalaBinaryVersion.value == "2.12" || scalaBinaryVersion.value == "2.13"),
```
We can remove this now since we won't support 2.11 any more, and `sqlDeltaImport` can support both Scala 2.12 and 2.13.

zsxwing commented on build.sbt:
```diff
@@ -616,22 +622,26 @@ lazy val sqlDeltaImport = (project in file("sql-delta-import"))
-      "org.apache.spark" % "spark-sql_2.12" % "3.0.0" % "provided",
-      "io.delta" % "delta-core_2.12" % "0.7.0" % "provided",
+      "io.netty" % "netty-buffer"  % "4.1.63.Final",
+      "org.apache.spark" % "spark-sql_2.13" % "3.2.0" % "provided",
```
```suggestion
      "org.apache.spark" %% "spark-sql" % "3.2.0" % "provided",
```
Could you also update the libraries below?

zsxwing commented on build.sbt:
```diff
@@ -616,22 +622,26 @@ lazy val sqlDeltaImport = (project in file("sql-delta-import"))
     )
   )
   .settings(releaseSettings)
+/*
```
Could you explain why you need to comment out flink? What's the issue you hit?

cheleb replied:
restored ... probably a dependency hell, solved now.

zsxwing commented on build.sbt:
```diff
@@ -40,9 +40,9 @@ val sparkVersion = "2.4.3"
 val parquet4sVersion = "1.2.1"
 val parquetHadoopVersion = "1.10.1"
-val scalaTestVersion = "3.0.5"
+val scalaTestVersion = "3.2.10"
```
Could you upgrade scala test to `3.0.8`? This should minimize the changes needed in this PR.

cheleb replied:
Symbol 'type org.scalatest.funsuite.AnyFunSuite' is missing from the classpath.
This symbol is required by 'class org.apache.spark.SparkFunSuite'.
Make sure that type AnyFunSuite is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
A full rebuild may help if 'SparkFunSuite.class' was compiled against an incompatible version of org.scalatest.funsuite.

zsxwing replied:
Not sure why you hit the issue. I changed this to 3.0.8 and reverted changes about AnyFunSuite locally and it worked.

I also checked SparkFunSuite and it doesn't use `org.scalatest.funsuite.AnyFunSuite`.

cheleb replied:
Cool, a bit early here, I will revert this asap, sorry for the noise.

cheleb replied:
Hum,  I'm confused CI does not complain, but locally I still have:

```
[error] /Users/cheleb/projects/delta.io/connectors/golden-tables/src/test/scala/io/delta/golden/GoldenTables.scala:56:43: Symbol 'type org.scalatest.funsuite.AnyFunSuite' is missing from the classpath.
[error] This symbol is required by 'class org.apache.spark.SparkFunSuite'.
[error] Make sure that type AnyFunSuite is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
[error] A full rebuild may help if 'SparkFunSuite.class' was compiled against an incompatible version of org.scalatest.funsuite.
[error] class GoldenTables extends QueryTest with SharedSparkSession {
[error]                                           ^
[error] one error found
```

zsxwing commented on build.sbt:
```diff
@@ -40,9 +40,9 @@ val sparkVersion = "2.4.3"
+val scalaTestVersion = "3.2.10"
 // Versions for Hive 3
-val hadoopVersion = "3.1.0"
+val hadoopVersion = "3.3.1"
```
Could you explain why we need to upgrade Hadoop version?

cheleb replied:
Mistake :-/

zsxwing commented on build.sbt:
```diff
@@ -52,10 +52,10 @@ val tezVersionForHive2 = "0.8.4"
   fork := true,
   javacOptions ++= Seq("-source", "1.8", "-target", "1.8", "-Xlint:unchecked"),
-  scalacOptions ++= Seq("-target:jvm-1.8", "-Ywarn-unused-import"),
+  scalacOptions ++= Seq("-target:jvm-1.8"),//, "-Ywarn-unused-import"),
```
Does `-Ywarn-unused-import` not work in Scala 2.13?

zsxwing commented on build.sbt:
```diff
@@ -577,13 +583,13 @@ lazy val compatibility = (project in file("oss-compatibility-tests"))
-      "org.apache.spark" % "spark-sql_2.12" % "3.1.1" % "test",
-      "io.delta" % "delta-core_2.12" % "1.0.0" % "test",
+      "org.scalatest" %% "scalatest" % "3.2.10" % "test",
+      "org.apache.spark" % "spark-sql_2.13" % "3.2.0" % "test",
```
It's better to use Scala 2.12 here since 2.13 is not GAed in Spark

zsxwing commented on build.sbt:
```diff
@@ -593,21 +599,21 @@ lazy val goldenTables = (project in file("golden-tables")) settings (
-    "org.apache.spark" % "spark-sql_2.12" % "3.0.0" % "test",
-    "io.delta" % "delta-core_2.12" % "0.8.0" % "test",
+    "org.scalatest" %% "scalatest" % scalaTestVersion % "test",
+    "org.apache.spark" % "spark-sql_2.13" % "3.2.0" % "test",
```
It's better to use Scala 2.12 here since 2.13 is not GAed in Spark

zsxwing reviewed (changes requested):
Thanks a lot for the PR. I saw there are multiple files created. Could you explain why we need to create shims? Will dropping Scala 2.11 support help? It would be great if you can point out which lines are different in these files so that that's much easier for reviewers to read.

cheleb commented:
> 

diff -r src/main/scala-2.12/io/delta/standalone/internal/ src/main/scala-2.13/io/delta/standalone/internal/

**Collection API changes:** 

io/delta/standalone/internal/BufferingLogDeletionIterator.scala 

2.12 <       filesToDelete.enqueue(maybeDeleteFiles: _*)
---
2.13 >       filesToDelete.enqueueAll(maybeDeleteFiles)

io/delta/standalone/internal/util/CaseInsensitiveMap.scala 

>   override def removed(key: String): Map[String, T] =
>      new CaseInsensitiveMap(originalMap.removed(key.toLowerCase(Locale.ROOT)))
> 
>   override def updated[V1 >: T](key: String, value: V1): Map[String, V1] =
>      new CaseInsensitiveMap(originalMap.updated(key.toLowerCase(Locale.ROOT), value))
> 
50,52d55
<   override def -(key: String): Map[String, T] = {
<     new CaseInsensitiveMap(originalMap.filter(!_._1.equalsIgnoreCase(key)))
<   }


**Parallel collection** are an external module in 2.13 and must be imported

io/delta/standalone/internal/SnapshotImpl.scala

2.13 > import scala.collection.parallel.CollectionConverters._

zsxwing commented:
> filesToDelete.enqueue(maybeDeleteFiles: _*)

We can replace it with `filesToDelete ++= maybeDeleteFiles` which is working in both Scala 2.12 and 2.13.

> io/delta/standalone/internal/util/CaseInsensitiveMap.scala

I will check whether we can avoid two copies of this.

> Parallel collection are an external module in 2.13 and must be imported

Could you just follow this pattern https://github.com/apache/spark/commit/7aca0dd658b8bda05574b3df3254aaf66eb2a174#diff-372e634ed9f323b52e037494e1c8e953486d8539965a8a1733fc267eaff2f765R79-R85

zsxwing commented:
Tried locally. Yep it seems we have to duplicate CaseInsensitiveMap.

By the way, we also need changes like this to disable mima for Scala 2.13: https://github.com/delta-io/delta/commit/c4227f3dd4c9bb174e51bfc5ee86855d4f23eda9#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91R222-R230

cheleb commented:
Hum, sorry about that !
Might be my IDE, that was misleading me.

On Wed, Jan 26, 2022 at 7:47 AM Shixiong Zhu ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In build.sbt
> <https://github.com/delta-io/connectors/pull/259#discussion_r792350290>:
>
> > @@ -40,9 +40,9 @@ val sparkVersion = "2.4.3"
>  val hiveDeltaVersion = "0.5.0"
>  val parquet4sVersion = "1.2.1"
>  val parquetHadoopVersion = "1.10.1"
> -val scalaTestVersion = "3.0.5"
> +val scalaTestVersion = "3.2.10"
>
> Not sure why you hit the issue. I changed this to 3.0.8 and reverted
> changes about AnyFunSuite locally and it worked.
>
> I also checked SparkFunSuite and it doesn't use
> org.scalatest.funsuite.AnyFunSuite.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/delta-io/connectors/pull/259#discussion_r792350290>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC4NK26Z3L7PNX4RTDO2R3UX6KH5ANCNFSM5L7SZVGQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
"Computers are useless. They can only give you answers."
- Pablo Picasso -

zsxwing commented on testStandaloneCosmetic/src/test/scala/ShadedJarSuite.scala:
```diff
@@ -16,6 +16,7 @@
 
 import org.scalatest.FunSuite
 
+
```
nit: unnecessary change

zsxwing commented on standalone/src/test/scala/io/delta/standalone/internal/ActionBuildersSuite.scala:
```diff
@@ -21,12 +21,11 @@ import java.util.{Collections, Optional}
 
-class ActionBuildersSuite extends FunSuite {
+
+class ActionBuildersSuite extends org.scalatest.FunSuite {
```
nit: unnecessary change

zsxwing commented on standalone/src/test/scala/io/delta/standalone/internal/ActionBuildersSuite.scala:
```diff
@@ -21,12 +21,11 @@ import java.util.{Collections, Optional}
 import scala.collection.JavaConverters._
 
-import org.scalatest.FunSuite
-
```
nit: unnecessary change

zsxwing commented on standalone/src/main/scala-2.12/io/delta/standalone/internal/util/CaseInsensitiveMap.scala:
```diff
@@ -37,7 +37,6 @@ private[internal] class CaseInsensitiveMap[T] private (val originalMap: Map[Stri
 
   override def contains(k: String): Boolean =
     keyLowerCasedMap.contains(k.toLowerCase(Locale.ROOT))
-
```
nit: unnecessary change

zsxwing commented on hive/src/main/scala/io/delta/hive/CaseInsensitiveMap.scala:
```diff
@@ -63,9 +63,12 @@ class CaseInsensitiveMap[T] private (val originalMap: Map[String, T])
-  override def -(key: String): Map[String, T] = {
-    new CaseInsensitiveMap(originalMap.filterKeys(!_.equalsIgnoreCase(key)))
-  }
+  override def removed(key: String): Map[String, T] =
```
We need to fork this file for Scala 2.12

zsxwing commented on hive-test/src/test/scala/io/delta/hive/test/HiveTest.scala:
```diff
@@ -30,7 +30,9 @@ import org.apache.hadoop.hive.conf.HiveConf
 import org.apache.hadoop.hive.ql.metadata.Hive
 import org.apache.hadoop.hive.ql.session.SessionState
-import org.scalatest.{BeforeAndAfterAll, FunSuite}
+import org.scalatest.BeforeAndAfterAll
+import org.scalatest.FunSuite
```
nit: unnecessary change

zsxwing commented on build.sbt:
```diff
@@ -661,3 +683,5 @@ lazy val flinkConnector = (project in file("flink-connector"))
   )
   .settings(skipReleaseSettings)
   .dependsOn(standaloneCosmetic % "provided")
+
+  
```
nit: unnecessary change

zsxwing commented on build.sbt:
```diff
@@ -53,13 +53,19 @@ val hadoopVersionForHive2 = "2.7.2"
 val tezVersionForHive2 = "0.8.4"
 
+def scalacWarningUnusedImport(version: String) = version match {
+    case v if v.startsWith("2.13.") =>
+    "-Ywarn-unused:imports"
```
nit: we can merge these two lines

zsxwing commented on .github/workflows/test.yaml:
```diff
@@ -6,7 +6,7 @@ jobs:
     strategy:
       matrix:
-        scala: [2.12.8, 2.11.12]
+        scala: [2.13.8, 2.12.8]
```
For this file, could you add `if: matrix.scala != '2.13.8'` for `Run Flink tests` since Flink doesn't support Scala 2.13?

zsxwing replied:
Could you comment out 
```
- name: Run sqlDeltaImport tests (Scala 2.12 only)
        run: build/sbt "++ ${{ matrix.scala }}" sqlDeltaImport/test
        if: matrix.scala != '2.11.12'
```
in this file to disable sqlDeltaImport tests? I noticed the tests are not working with Spark 3.2.0. I will make a followup PR to fix the tests.

zsxwing commented on build.sbt:
```diff
@@ -160,7 +166,7 @@ lazy val hive = (project in file("hive")) dependsOn(standaloneCosmetic) settings
   libraryDependencies ++= Seq(
     "org.apache.hadoop" % "hadoop-client" % hadoopVersion % "provided",
-    "org.apache.parquet" % "parquet-hadoop" % "1.10.1" % "provided",
+    "org.apache.parquet" % "parquet-hadoop" % "1.12.2" % "provided",
```
unnecessary change

zsxwing commented:
Hm, we also need to bump jackson and json4s version. You can find the proper version in 
https://github.com/apache/spark/blob/v3.2.0/pom.xml#L170 and https://github.com/apache/spark/blob/v3.2.0/pom.xml#L916

zsxwing commented:
@cheleb I realized we still need to support Scala 2.11. I opened https://github.com/cheleb/connectors/pull/1 to add one commit to your branch. Could you take a look?

allisonport-db commented:
Could we also update [run_examples.py](https://github.com/delta-io/connectors/blob/master/examples/run_examples.py#L123) while we're at it

zsxwing commented on build.sbt:
```diff
@@ -53,13 +54,18 @@ val hadoopVersionForHive2 = "2.7.2"
   organization := "io.delta",
-  scalaVersion := scala212,
-  crossScalaVersions := Seq(scala212, scala211),
+  scalaVersion := scala213,
```
Could you change this back to scala212 as Scala 2.12 is the only version that all projects support?

zsxwing reviewed (approved):
LGTM.

Thanks a lot for your contribution! I'm going to merge this PR as it's pretty large now. Could you address the following comment in a separate PR?

> Could we also update [run_examples.py](https://github.com/delta-io/connectors/blob/master/examples/run_examples.py#L123) while we're at it?
